zip-0032.rst

```diff
@@ -0,0 +1,464 @@
+least nonnegative integer yielding a valid diversifier.
+
+
+Specification: Sprout key derivation
+====================================
+
+For completeness, we define a system for deriving a tree of Sprout key components. It is unlikely that this
+will garner much usage once Sapling activates, but is presented for those users who may require it.
+
+Sprout extended keys
+--------------------
+
+Due to the way Sprout keys are constructed and used, it is not possible to derive incoming viewing keys or
+payment addresses in parallel with spending keys. Nor is it possible to implement non-hardened derivation.
+We therefore only define and derive Sprout extended spending keys.
+
+We represent a Sprout extended spending key as (*a*\ :sub:`sk`\ , *c*), where *a*\ :sub:`sk` is the normal
+Sprout spending key, and *c* is the chain code.
+
+Helper functions
+----------------
+
+Let EncodeASK(*a*\ :sub:`sk`) be the 32-byte encoding of *a*\ :sub:`sk` in the raw encoding of a Sprout
+spending key (excluding lead bytes) as specified in [#sapling-spec]_ section 5.6.8.
+
+Let DecodeASK(*ASK*) be the result of clearing the 4 most significant bits of the first byte of *ASK*,
+and decoding the 32-byte result according to the inverse of EncodeASK.
+
+Master key generation
+---------------------
+
+Let *S* be a seed byte sequence of a chosen length, which MUST be at least 32 bytes.
+
+- Calculate *I* = BLAKE2b-512("ZcashIP32_Sprout", *S*).
+- Split *I* into two 32-byte sequences, I\ :sub:`L` and I\ :sub:`R`\ .
+- Use DecodeASK(*I*\ :sub:`L`\ ) as the master spending key a\ :sub:`sk,m`\ .
+- Use *I*\ :sub:`R` as the master chain code *c*\ :sub:`m`\ .
+
+Child key derivation
+--------------------
+
+CDKsk((*a*\ :sub:`sk,par`\ , *c*\ :sub:`par`\ ), *i*) → (*a*\ :sub:`sk,i`\ , *c*\ :sub:`i`\ )
+
+- Check whether *i* ≥ 2\ :sup:`31` (whether the child is a hardened key).
+
+  - If so (hardened child): let *I* = PRF\ :sup:`expand`\ (*c*\ :sub:`par`\ , [0x80] || EncodeASK(*a*\ :sub:`sk,par`\ ) || I2LEOSP\ :sub:`32`\ (*i*))
+  - If not (normal child): return failure
+
+- Split *I* into two 32-byte sequences, *I*\ :sub:`L` and *I*\ :sub:`R`\ .
+- Use DecodeASK(*I*\ :sub:`L`\ ) as the child spending key a\ :sub:`sk,i`\ .
+- Use *I*\ :sub:`R` as the child chain code *c*\ :sub:`i`\ .
+
+
+Specification: Wallet usage
+===========================
+
+Existing Zcash-supporting HD wallets all use BIP 44 [#bip-0044]_ to organize their derived keys. In order to
+more easily mesh with existing user experiences, we broadly follow BIP 44's design here. However, we have
+altered the design where it makes sense to leverage features of shielded addresses.
+
+Key path levels
+---------------
+
+Both Sprout and Sapling key paths have the following three path levels at the top, all of which use hardened
+derivation:
+
+- ``purpose``: a constant set to 32' (or 0x80000020) following the BIP 43 recommendation. It indicates that
```

Comment:
explain that `i' = i+2^31`

Reply:
Added to the existing hardened notation definitions up the top.